Daos 2759 test: Avocado functional tests do not run in DAOS env installed from RPMs

GitHuaKuang opened the pull request:
1.  Renamed install/lib that was holding libraries to lib64
     2.  Copied src/tests/ftest to $PREFIX/lib/daos/TESTING/ftest
     3.  Copied .build_vars.* to install/lib to allow existing test code run
     4.  Updated some avocado test setup code to adapt the new file locations
     5.  Updated rpms/daos.spec to use the new code structure for packaging
     6.  Updated CI unit test scripts to use the new library path
     7.  Moved CI function tests to run from the new directory
     8. Copied src/utils/py to $PREFIX/lib/daos/utils/py

daosbuild1 commented on src/tests/ftest/util/server_utils.py:
```diff
@@ -77,15 +77,18 @@ def create_server_yaml(basepath):
     """
+    with open(os.path.join(basepath, ".build_vars.json")) as json_vars:
+            build_vars = json.load(json_vars)
+    
```
(style)  trailing whitespace
(pylint-trailing-whitespace) Trailing whitespace

daosbuild1 commented on src/tests/ftest/util/server_utils.py:
```diff
@@ -77,15 +77,18 @@ def create_server_yaml(basepath):
 
     """
+    with open(os.path.join(basepath, ".build_vars.json")) as json_vars:
+            build_vars = json.load(json_vars)
```
(pylint-bad-indentation) Bad indentation. Found 12 spaces, expected 8

daosbuild1 commented on src/tests/ftest/util/server_utils.py:
```diff
@@ -204,9 +207,9 @@ def run_server(hostfile, setname, basepath, uri_path=None, env_dict=None,
         server_cmd.extend([daos_srv_bin,
                            "--debug",
-                           "--config", '{}/{}'.format(basepath, AVOCADO_FILE),
+                           "--config", '{}/{}'.format(build_vars["PREFIX"], AVOCADO_FILE),
```
(pylint-line-too-long) Line too long (90/80)

phender replied:
```suggestion
                           "--config", os.path.realpath(AVOCADO_FILE),
```

daosbuild1 commented on src/tests/ftest/util/server_utils.py:
```diff
@@ -77,15 +77,18 @@ def create_server_yaml(basepath):
     # Read the baseline conf file data/daos_server_baseline.yml
     try:
-        with open('{}/{}'.format(basepath, DEFAULT_FILE), 'r') as read_file:
+        with open('{}/{}'.format(build_vars["PREFIX"], DEFAULT_FILE), 'r') as read_file:
```
(pylint-line-too-long) Line too long (88/80)

phender replied:
If using relative paths:
```suggestion
        with open(default_yaml, 'r') as read_file:
```

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1048/1/execution/node/47/log

daosbuild1 commented:
Test stage Build on CentOS 7 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1048/1/execution/node/203/log

daosbuild1 commented:
Test stage Build on Leap 42.3 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1048/1/execution/node/224/log

daosbuild1 commented:
Test stage Build on Ubuntu 18.04 with Clang completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1048/1/execution/node/221/log

daosbuild1 commented:
Test stage Build on SLES 12.3 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1048/1/execution/node/237/log

daosbuild1 commented:
Test stage Build on Leap 15 with Intel-C completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1048/1/execution/node/236/log

daosbuild1 commented:
Test stage Build RPM on CentOS 7 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1048/1/execution/node/233/log

daosbuild1 commented:
Test stage Build RPM on SLES 12.3 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1048/1/execution/node/386/log

phender commented on src/tests/ftest/util/server_utils.py:
```diff
@@ -41,8 +41,8 @@
 
-DEFAULT_FILE = "src/tests/ftest/data/daos_server_baseline.yaml"
-AVOCADO_FILE = "src/tests/ftest/data/daos_avocado_test.yaml"
+DEFAULT_FILE = "lib/daos/TESTING/ftest/data/daos_server_baseline.yaml"
```
I'd suggest using relative paths here:
1) git move the `src/tests/ftest/data` directory to `src/tests/ftest/util/data`
2) Use these definitions for the constants:
```
DEFAULT_FILE = "./data/daos_server_baseline.yaml"
AVOCADO_FILE = "./data/daos_avocado_test.yaml"
```

phender commented on src/tests/ftest/util/server_utils.py:
```diff
@@ -77,15 +77,18 @@ def create_server_yaml(basepath):
         traceback.print_exception(excpn.__class__, excpn, sys.exc_info()[2])
         raise ServerFailed(
-            "Failed to Read {}/{}".format(basepath, DEFAULT_FILE))
+            "Failed to Read {}/{}".format(build_vars["PREFIX"], DEFAULT_FILE))
```
```suggestion
            "Failed to Read {}".format(default_yaml))
```

phender commented on src/tests/ftest/util/server_utils.py:
```diff
@@ -121,13 +124,13 @@ def create_server_yaml(basepath):
     # This will be used to start with daos_server -o option.
     try:
-        with open('{}/{}'.format(basepath,
+        with open('{}/{}'.format(build_vars["PREFIX"],
```
```suggestion
        with open(
```

phender commented on src/tests/ftest/util/server_utils.py:
```diff
@@ -121,13 +124,13 @@ def create_server_yaml(basepath):
         print("<SERVER> Exception occurred: {0}".format(str(excpn)))
         traceback.print_exception(excpn.__class__, excpn, sys.exc_info()[2])
-        raise ServerFailed("Failed to Write {}/{}".format(basepath,
+        raise ServerFailed("Failed to Write {}/{}".format(build_vars["PREFIX"],
```
```suggestion
        raise ServerFailed("Failed to Write {}".format(
```

brianjmurrell commented on src/tests/SConscript:
```diff
@@ -31,6 +45,14 @@ def scons():
+    spath = join(denv.subst("$PREFIX"), "../src/tests/ftest/")
+    # denv.InstallAs('$PREFIX/lib/daos/TESTING/ftest/', spath)
+    if not check_dir_exists(tpath):
+        os.symlink(spath, tpath)
```
You cannot do this.  The point of `$PREFIX` (`install/` in a local installation; `/usr` in an RPM based installation) is that it supposed to be a self-contained tree of everything needed to run the tests.  You should be able to remove everything above it and it still works.

But do think of the RPM installation case.  `/usr/lib/daos/TESTING/ftest` will be a symlink to where?  The local source tree will not exist on the machine that the RPM is installed on.

GitHuaKuang replied:
Not clear how RPM is built, but this src/tests directory is not available in RPM pkg, so this SConscript will not be executed, right?

GitHuaKuang replied:
changed to copy instead of symlink.

brianjmurrell commented on src/tests/ftest/util/server_utils.py:
```diff
@@ -41,8 +41,8 @@
-DEFAULT_FILE = "src/tests/ftest/data/daos_server_baseline.yaml"
-AVOCADO_FILE = "src/tests/ftest/data/daos_avocado_test.yaml"
+DEFAULT_FILE = "lib/daos/TESTING/ftest/data/daos_server_baseline.yaml"
+AVOCADO_FILE = "lib/daos/TESTING/ftest/data/daos_avocado_test.yaml"
```
This should migrate more towards where the server config file will be in a production environment -- probably $PREFIX/etc/daos, for example.  That will make it port to an RPM installation more easily.

GitHuaKuang replied:
talking Phil's suggestion to use relative path.

GitHuaKuang replied:
changed to $PREFIX/etc/daos

mjean308 commented on src/tests/ftest/util/server_utils.py:
```diff
@@ -204,9 +207,9 @@ def run_server(hostfile, setname, basepath, uri_path=None, env_dict=None,
                            "start", "-i",
-                           "-a", os.path.join(basepath, "install", "tmp")])
+                           "-a", os.path.join(build_vars["PREFIX"], "tmp")])
 
```
I would not use build_vars["PREFIX"] to locate tmp.    Currently we need this to be a shared directory and in rpms this is not.  (/usr/tmp)     If this could  be an ENV then the user would be able to create their own shared dir.    
This will not be necessary once we have systemctl for daos_server but it will not work for rpms as written right now.

brianjmurrell replied:
Actually `build_vars["PREFIX"]` is still the shared directory in CI.

But the point of being able to set that to some other value in environments outside of our CI is important.

GitHuaKuang replied:
So we can used build_vars["TMPDIR"]? one question is what's the default value if TMPDIR is not set by user? and which file should this env be added to?

brianjmurrell replied:
`build_vars["TMPDIR"]` doesn't exist.

GitHuaKuang replied:
right, either we add TMPDIR in now if we want to solve it in this PR, or leave the user configured tmp dir to another later ticket.

brianjmurrell replied:
So, _TMPDIR_ is not really what this is.  It's "SHAREDDIR" or something like that.  For CI testing, it can be `build_vars["PREFIX"]`.  That will be a shared dir.  But we need to also support a mechanism for non-CI users to set this dir.  Maybe non-CI users just edit the build_vars.json to point at their shared dir.  Maybe some other mechanism.

GitHuaKuang replied:
Tried to add a TMPDIR to daos/.build_vars.json:
  "PREFIX": "/home/huakuang/daos/install", 
  "TMPDIR": "/home/huakuang",

Then modified use it in server_utils.py here:
                           os.path.join(build_vars["TMPDIR"], "tmp")])

Launch tests from install/lib/daos/TESTING/ftest worked fine.
 ./launch.py -c -a -r -ts boro-58 conthandle

GitHuaKuang replied:
Allow user to set TMPDIR, if  not set, default it to PREFIX

daosbuild1 commented on src/tests/SConscript:
```diff
@@ -1,5 +1,7 @@
 """Build tests"""
 import daos_build
+import os
+from os.path import join, isdir
```
(pylint-unused-import) Unused isdir imported from os.path

daosbuild1 replied:
(pylint-unused-import) Unused isdir imported from os.path

daosbuild1 commented on src/tests/ftest/util/server_utils.py:
```diff
@@ -41,9 +41,10 @@
-
+DEFAULT_FILE = "./data/daos_server_baseline.yaml"
+AVOCADO_FILE = "./data/daos_avocado_test.yaml"
+default_yaml = os.path.realpath(DEFAULT_FILE)
```
(pylint-invalid-name) Invalid constant name "default_yaml"

daosbuild1 replied:
(pylint-invalid-name) Invalid constant name "default_yaml"

daosbuild1 commented on src/tests/ftest/util/server_utils.py:
```diff
@@ -41,9 +41,10 @@
+DEFAULT_FILE = "./data/daos_server_baseline.yaml"
+AVOCADO_FILE = "./data/daos_avocado_test.yaml"
+default_yaml = os.path.realpath(DEFAULT_FILE)
+avocado_yaml = os.path.realpath(AVOCADO_FILE)
```
(pylint-invalid-name) Invalid constant name "avocado_yaml"

daosbuild1 replied:
(pylint-invalid-name) Invalid constant name "avocado_yaml"

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1048/2/execution/node/47/log

daosbuild1 commented:
Test stage Build RPM on CentOS 7 completed with status UNSTABLE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1048/2/execution/node/203/log

brianjmurrell commented on src/tests/ftest/util/server_utils.py:
```diff
@@ -127,14 +124,12 @@ def create_server_yaml(basepath, log_filename):
     try:
-        with open('{}/{}'.format(build_vars["PREFIX"],
-                                 AVOCADO_FILE), 'w') as write_file:
+        with open(avocado_yaml, 'w') as write_file:
```
IIUC, this file will be under `$PREFIX/lib/daos/...`, is that correct?  If so, this file is not going to be writable in an RPM based environment.  It should be in `$PREFIX/etc/daos/` probably.

daosbuild1 commented:
Test stage Build RPM on SLES 12.3 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1048/2/execution/node/243/log

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1048/3/execution/node/47/log

daosbuild1 commented:
Test stage Build RPM on CentOS 7 completed with status UNSTABLE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1048/3/execution/node/228/log

daosbuild1 commented:
Test stage Build RPM on SLES 12.3 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1048/3/execution/node/243/log

daosbuild1 commented on src/tests/SConscript:
```diff
@@ -31,6 +33,14 @@ def scons():
 
+    # install yamls files
+    env.Install('$PREFIX/etc/daos', ['ftest/data/daos_server_baseline.yaml',
+        'ftest/data/daos_avocado_test.yaml'])
```
(pylint-bad-continuation) Wrong continued indentation (add 29 spaces).

daosbuild1 replied:
(pylint-bad-continuation) Wrong continued indentation (add 29 spaces).

daosbuild1 commented on src/tests/ftest/util/server_utils.py:
```diff
@@ -206,9 +210,9 @@ def run_server(hostfile, setname, basepath, uri_path=None, env_dict=None,
         server_cmd.extend([daos_srv_bin,
                            "--debug",
-                           "--config", '{}/{}'.format(basepath, AVOCADO_FILE),
+                           "--config", '{}/{}'.format(build_vars["PREFIX"], AVOCADO_FILE),
```
(pylint-line-too-long) Line too long (90/80)

daosbuild1 replied:
(pylint-line-too-long) Line too long (90/80)

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1048/4/execution/node/47/log

daosbuild1 commented:
Test stage Build RPM on CentOS 7 completed with status UNSTABLE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1048/4/execution/node/222/log

daosbuild1 commented:
Test stage Build RPM on SLES 12.3 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1048/4/execution/node/243/log

daosbuild1 commented:
Test stage Build on Leap 42.3 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1048/4/execution/node/206/log

daosbuild1 commented:
Test stage Build on SLES 12.3 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1048/4/execution/node/219/log

daosbuild1 commented:
Test stage Build on CentOS 7 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1048/4/execution/node/233/log

daosbuild1 commented:
Test stage Build on Ubuntu 18.04 with Clang completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1048/4/execution/node/225/log

daosbuild1 commented:
Test stage Build on Leap 15 with Intel-C completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1048/4/execution/node/218/log

daosbuild1 commented on src/tests/ftest/container/global_handle.py:
```diff
@@ -166,7 +158,7 @@ def test_global_handle(self):
             # for now verifying global handle in the same process which is not
             # the intended use case
-            check_handle(sct_pool_handle, sct_pool_uuid, sct_cont_handle, 0)
+            check_handle(self, sct_pool_handle, sct_pool_uuid, sct_cont_handle, 0)
```
(pylint-line-too-long) Line too long (82/80)

daosbuild1 commented on src/tests/ftest/pool/global_handle.py:
```diff
@@ -32,31 +32,24 @@
 
         # note that the handle is stored inside the pool as well
-        dummy_local_handle = pool.global2local(context, iov_len, buf_len, buf)
+        dummy_local_handle = pool.global2local(self.context, iov_len, buf_len, buf)
```
(pylint-line-too-long) Line too long (83/80)

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1048/5/execution/node/47/log

daosbuild1 commented:
Test stage Build RPM on CentOS 7 completed with status UNSTABLE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1048/5/execution/node/213/log

daosbuild1 commented:
Test stage Build RPM on SLES 12.3 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1048/5/execution/node/243/log

daosbuild1 commented:
Test stage Build on SLES 12.3 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1048/5/execution/node/225/log

daosbuild1 commented:
Test stage Build on Leap 42.3 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1048/5/execution/node/224/log

daosbuild1 commented:
Test stage Build on CentOS 7 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1048/5/execution/node/233/log

daosbuild1 commented:
Test stage Build on Ubuntu 18.04 with Clang completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1048/5/execution/node/223/log

daosbuild1 commented:
Test stage Build on Leap 15 with Intel-C completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1048/5/execution/node/216/log

daosbuild1 commented on src/control/SConscript:
```diff
@@ -119,15 +120,16 @@ def scons():
-    if not conf.CheckGoVersion():
-        print 'no usable Go compiler found (yum install golang?)'
+    conf = Configure(denv, custom_tests={'check_go_version': check_go_version})
+    if not conf.check_go_version():
```
(pylint-no-member) Instance of 'Configure' has no 'check_go_version' member

daosbuild1 commented on src/tests/ftest/util/server_utils.py:
```diff
@@ -206,9 +210,10 @@ def run_server(hostfile, setname, basepath, uri_path=None, env_dict=None,
                            "--debug",
-                           "--config", '{}/{}'.format(basepath, AVOCADO_FILE),
+                           "--config", '{}/{}'.format(build_vars["PREFIX"],
+			   AVOCADO_FILE),
```
(pylint-bad-continuation) Wrong continued indentation (add 48 spaces).

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1048/6/execution/node/47/log

daosbuild1 commented:
Test stage Build RPM on CentOS 7 completed with status UNSTABLE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1048/6/execution/node/218/log

daosbuild1 commented:
Test stage Build RPM on SLES 12.3 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1048/6/execution/node/243/log

daosbuild1 commented:
Test stage Build on Leap 42.3 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1048/6/execution/node/221/log

daosbuild1 commented:
Test stage Build on SLES 12.3 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1048/6/execution/node/222/log

daosbuild1 commented:
Test stage Build on CentOS 7 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1048/6/execution/node/233/log

daosbuild1 commented:
Test stage Build on Ubuntu 18.04 with Clang completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1048/6/execution/node/220/log

daosbuild1 commented:
Test stage Build on Leap 15 with Intel-C completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1048/6/execution/node/225/log

daosbuild1 commented on src/tests/ftest/util/server_utils.py:
```diff
@@ -204,11 +208,10 @@ def run_server(hostfile, setname, basepath, uri_path=None, env_dict=None,
-                           "start", "-i",
-                           "-a", os.path.join(basepath, "install", "tmp")])
+        server_cmd.extend([daos_srv_bin, "--debug", "--config",
+                          '{}/{}'.format(build_vars["PREFIX"], AVOCADO_FILE),
```
(pylint-bad-continuation) Wrong continued indentation (add 1 space).

daosbuild1 commented on src/tests/ftest/util/server_utils.py:
```diff
@@ -204,11 +208,10 @@ def run_server(hostfile, setname, basepath, uri_path=None, env_dict=None,
-                           "-a", os.path.join(basepath, "install", "tmp")])
+        server_cmd.extend([daos_srv_bin, "--debug", "--config",
+                          '{}/{}'.format(build_vars["PREFIX"], AVOCADO_FILE),
+                          "start", "-i", "-a",
```
(pylint-bad-continuation) Wrong continued indentation (add 1 space).

daosbuild1 commented on src/tests/ftest/util/server_utils.py:
```diff
@@ -204,11 +208,10 @@ def run_server(hostfile, setname, basepath, uri_path=None, env_dict=None,
+        server_cmd.extend([daos_srv_bin, "--debug", "--config",
+                          '{}/{}'.format(build_vars["PREFIX"], AVOCADO_FILE),
+                          "start", "-i", "-a",
+                          os.path.join(build_vars["PREFIX"], "tmp")])
```
(pylint-bad-continuation) Wrong continued indentation (add 1 space).

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1048/7/execution/node/47/log

daosbuild1 commented:
Test stage Build RPM on CentOS 7 completed with status UNSTABLE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1048/7/execution/node/208/log

daosbuild1 commented:
Test stage Build RPM on SLES 12.3 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1048/7/execution/node/243/log

daosbuild1 commented:
Test stage Build on CentOS 7 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1048/7/execution/node/180/log

daosbuild1 commented:
Test stage Build on Leap 42.3 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1048/7/execution/node/232/log

daosbuild1 commented:
Test stage Build on SLES 12.3 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1048/7/execution/node/233/log

daosbuild1 commented:
Test stage Build on Leap 15 with Intel-C completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1048/7/execution/node/227/log

daosbuild1 commented:
Test stage Build on Ubuntu 18.04 with Clang completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1048/7/execution/node/220/log

daosbuild1 commented:
Test stage Build RPM on CentOS 7 completed with status UNSTABLE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1048/8/execution/node/222/log

daosbuild1 commented:
Test stage Build RPM on SLES 12.3 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1048/8/execution/node/243/log

daosbuild1 commented:
Test stage Build on Leap 42.3 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1048/8/execution/node/216/log

daosbuild1 commented:
Test stage Build on SLES 12.3 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1048/8/execution/node/205/log

daosbuild1 commented:
Test stage Build on CentOS 7 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1048/8/execution/node/233/log

daosbuild1 commented:
Test stage Build on Ubuntu 18.04 with Clang completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1048/8/execution/node/225/log

daosbuild1 commented:
Test stage Build on Leap 15 with Intel-C completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1048/8/execution/node/217/log

daosbuild1 commented:
Test stage Build RPM on CentOS 7 completed with status UNSTABLE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1048/9/execution/node/224/log

daosbuild1 commented:
Test stage Build RPM on SLES 12.3 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1048/9/execution/node/243/log

daosbuild1 commented on src/tests/SConscript:
```diff
@@ -1,5 +1,7 @@
 """Build tests"""
 import daos_build
+import os
+from os.path import join
```
(pylint-unused-import) Unused join imported from os.path

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1048/10/execution/node/47/log

daosbuild1 commented:
Test stage Build RPM on CentOS 7 completed with status UNSTABLE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1048/10/execution/node/223/log

daosbuild1 commented:
Test stage Build RPM on SLES 12.3 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1048/10/execution/node/243/log

daosbuild1 commented:
Test stage Build RPM on CentOS 7 completed with status UNSTABLE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1048/11/execution/node/222/log

daosbuild1 commented:
Test stage Build RPM on SLES 12.3 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1048/11/execution/node/243/log

daosbuild1 commented:
Test stage Build RPM on CentOS 7 completed with status UNSTABLE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1048/12/execution/node/214/log

daosbuild1 commented:
Test stage Build RPM on SLES 12.3 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1048/12/execution/node/244/log

phender commented on src/tests/ftest/container/global_handle.py:
```diff
@@ -26,32 +26,23 @@
 from daos_cref import IOV
 
-def check_handle(pool_glob_handle, uuidstr, cont_glob_handle, rank):
+def check_handle(self, pool_glob_handle, uuidstr, cont_glob_handle, rank):
```
This function needs to move an become a method of the `GlobalHandle` class

phender commented on src/tests/ftest/pool/global_handle.py:
```diff
@@ -25,38 +25,31 @@
+from daos_api import DaosPool, DaosContainer, DaosApiError
 
-def check_handle(buf_len, iov_len, buf, uuidstr, rank):
+def check_handle(self, buf_len, iov_len, buf, uuidstr, rank):
```
This function needs to move an become a method of the `GlobalHandle` class

phender replied:
This function still needs to be defined inside of the `GlobalHandle` class, as it is currently defined outside of the class.

Also, to make use of the existing `TestWithServers.tearDown()` functionality the local `pool` variable should be `self.pool`.

daosbuild1 commented on utils/rpms/daos.spec:
```diff
@@ -135,7 +135,8 @@ sed -i -e s/${BUILDROOT//\//\\/}[^\"]\*/${PREFIX//\//\\/}/g %{?buildroot}%{_pref
 #mv %{?buildroot}/{usr/,}etc
 mkdir -p %{?buildroot}/%{_exec_prefix}/lib/%{name}
-mv %{?buildroot}%{_prefix}/lib64/daos %{?buildroot}/%{_exec_prefix}/lib/
+cp -al VERSION %{?buildroot}/%{_exec_prefix}/lib/%{name} 
```
(style)  trailing whitespace

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1048/14/execution/node/47/log

daosbuild1 commented:
Test stage Build RPM on CentOS 7 completed with status UNSTABLE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1048/15/execution/node/224/log

daosbuild1 commented:
Test stage Build RPM on SLES 12.3 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1048/15/execution/node/244/log

daosbuild1 commented:
Test stage Build RPM on CentOS 7 completed with status UNSTABLE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1048/16/execution/node/225/log

daosbuild1 commented:
Test stage Build RPM on SLES 12.3 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1048/16/execution/node/243/log

daosbuild1 commented:
Test stage Build RPM on CentOS 7 completed with status UNSTABLE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1048/17/execution/node/178/log

daosbuild1 commented:
Test stage Build RPM on SLES 12.3 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1048/17/execution/node/243/log

daosbuild1 commented:
Test stage Build RPM on CentOS 7 completed with status UNSTABLE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1048/18/execution/node/219/log

daosbuild1 commented:
Test stage Build RPM on SLES 12.3 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1048/18/execution/node/243/log

daosbuild1 commented:
Test stage Build RPM on CentOS 7 completed with status UNSTABLE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1048/19/execution/node/222/log

daosbuild1 commented:
Test stage Build RPM on SLES 12.3 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1048/19/execution/node/244/log

daosbuild1 commented:
Test stage Build RPM on CentOS 7 completed with status UNSTABLE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1048/20/execution/node/222/log

daosbuild1 commented:
Test stage Build RPM on SLES 12.3 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1048/20/execution/node/243/log

daosbuild1 commented:
Test stage Build RPM on CentOS 7 completed with status UNSTABLE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1048/21/execution/node/225/log

daosbuild1 commented:
Test stage Build RPM on SLES 12.3 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1048/21/execution/node/243/log

daosbuild1 commented on utils/certs/SConscript:
```diff
@@ -5,7 +5,7 @@ def scons():
     Import('env')
 
-    env.Install("$PREFIX/lib/daos/certgen", ['admin.cnf', 'agent.cnf', 'ca.cnf',
+    env.Install("$PREFIX/lib64/daos/certgen", ['admin.cnf', 'agent.cnf', 'ca.cnf',
```
(pylint-line-too-long) Line too long (82/80)

daosbuild1 replied:
(pylint-line-too-long) Line too long (82/80)

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1048/22/execution/node/47/log

daosbuild1 commented:
Test stage Build RPM on CentOS 7 completed with status UNSTABLE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1048/22/execution/node/224/log

daosbuild1 commented:
Test stage Build RPM on SLES 12.3 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1048/22/execution/node/235/log

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1048/23/execution/node/48/log

daosbuild1 commented:
Test stage Build RPM on CentOS 7 completed with status UNSTABLE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1048/23/execution/node/215/log

daosbuild1 commented:
Test stage Build RPM on SLES 12.3 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1048/23/execution/node/236/log

jolivier23 commented on src/tests/SConscript:
```diff
@@ -31,6 +32,13 @@ def scons():
+
+    # copy src/tests/ftest to $PREFIX/lib/daos/TESTING/ftest
+    spath = os.path.join(Dir('ftest').srcnode())
+    denv.InstallAs('$PREFIX/lib/daos/TESTING/ftest/', spath)
```
lib64?

GitHuaKuang replied:
It's actually /lib, not /lib64. We are moveing ftest to lib/daos/TESTING after moving all libraries under /lib to /lib64, this way, it's more in sync with RPM package code structure.

jolivier23 commented on utils/rpms/daos.spec:
```diff
@@ -132,13 +132,14 @@ scons %{?no_smp_mflags}              \
-mv %{?buildroot}%{_prefix}/lib{,64}
+#mv %{?buildroot}%{_prefix}/lib{,64}
 #mv %{?buildroot}/{usr/,}etc
 mkdir -p %{?buildroot}/%{_exec_prefix}/lib/%{name}
```
lib64?

GitHuaKuang replied:
yes.

jolivier23 commented on utils/rpms/daos.spec:
```diff
@@ -132,13 +132,14 @@ scons %{?no_smp_mflags}              \
-cp -al ftest.sh src/tests/ftest %{?buildroot}%{daoshome}/TESTING
-find %{?buildroot}%{daoshome}/TESTING/ftest -name \*.py[co] -print0 | xargs -r0 rm -f
+#cp -al VERSION %{?buildroot}/%{_exec_prefix}/lib/%{name}
+#mkdir -p %{?buildroot}/%{_exec_prefix}/lib/%{name}/certgen
```
lib64?

GitHuaKuang replied:
yes. fixed.

daosbuild1 commented on utils/certs/SConscript:
```diff
@@ -5,8 +5,9 @@ def scons():
-                                             'server.cnf', 'gen_certificates.sh'])
+    env.Install("$PREFIX/lib64/daos/certgen",
+                ['admin.cnf', 'agent.cnf', 'ca.cnf',
+                'server.cnf', 'gen_certificates.sh'])
```
(pylint-bad-continuation) Wrong continued indentation (add 1 space).

daosbuild1 replied:
(pylint-bad-continuation) Wrong continued indentation (add 1 space).

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1048/24/execution/node/47/log

daosbuild1 commented:
Test stage Build RPM on CentOS 7 completed with status UNSTABLE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1048/24/execution/node/219/log

daosbuild1 commented:
Test stage Build RPM on SLES 12.3 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1048/24/execution/node/235/log

daosbuild1 commented on SConstruct:
```diff
@@ -277,9 +277,12 @@ def scons():
-    env.InstallAs("$PREFIX/TESTING/.build_vars.json", ".build_vars.json")
-    env.InstallAs("$PREFIX/lib/daos/VERSION", "VERSION")
+    env.InstallAs("$PREFIX/lib/daos/TESTING/.build_vars.sh", ".build_vars.sh")
+    env.InstallAs("$PREFIX/lib/daos/TESTING/.build_vars.json", ".build_vars.json")
```
(pylint-line-too-long) Line too long (82/80)

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1048/25/execution/node/47/log

daosbuild1 commented:
Test stage Build RPM on CentOS 7 completed with status UNSTABLE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1048/25/execution/node/211/log

daosbuild1 commented:
Test stage Build RPM on SLES 12.3 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1048/25/execution/node/235/log

daosbuild1 commented on utils/certs/SConscript:
```diff
@@ -5,8 +5,9 @@ def scons():
-    env.Install("$PREFIX/lib/daos/certgen", ['admin.cnf', 'agent.cnf', 'ca.cnf',
-                                             'server.cnf', 'gen_certificates.sh'])
+    env.Install("$PREFIX/lib64/daos/certgen",
+                 ['admin.cnf', 'agent.cnf', 'ca.cnf',
```
(pylint-bad-continuation) Wrong continued indentation (remove 1 space).

daosbuild1 commented on utils/certs/SConscript:
```diff
@@ -5,8 +5,9 @@ def scons():
-                                             'server.cnf', 'gen_certificates.sh'])
+    env.Install("$PREFIX/lib64/daos/certgen",
+                 ['admin.cnf', 'agent.cnf', 'ca.cnf',
+                 'server.cnf', 'gen_certificates.sh'])
```
(pylint-bad-continuation) Wrong continued indentation (add 1 space).

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1048/27/execution/node/48/log

daosbuild1 commented:
Test stage Build RPM on CentOS 7 completed with status UNSTABLE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1048/27/execution/node/226/log

daosbuild1 commented:
Test stage run_test.sh completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1048/27/execution/node/423/log

daosbuild1 commented on utils/run_test.sh:
```diff
@@ -98,9 +98,9 @@ if [ -d "/mnt/daos" ]; then
     run_test src/rdb/raft_tests/raft_tests.py
     # Satisfy CGO requirements for go-spdk binding and internal daos imports
-    LD_LIBRARY_PATH="${SL_PREFIX}/lib:${SL_SPDK_PREFIX}/lib:${LD_LIBRARY_PATH}"
+    LD_LIBRARY_PATH="${SL_PREFIX}/lib64:${SL_SPDK_PREFIX}/lib64:${LD_LIBRARY_PATH}"
```
(style)  line over 80 characters

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1048/28/execution/node/47/log

daosbuild1 commented:
Test stage Build RPM on CentOS 7 completed with status UNSTABLE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1048/28/execution/node/177/log

daosbuild1 commented:
Test stage Test CentOS 7 RPMs completed with status FAILURE.  https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-1048/28/display/redirect

daosbuild1 commented:
Test stage Functional completed with status FAILURE.  https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-1048/28/display/redirect

daosbuild1 commented:
Test stage Functional_Hardware completed with status FAILURE.  https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-1048/28/display/redirect

daosbuild1 commented:
Test stage Build RPM on CentOS 7 completed with status UNSTABLE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1048/29/execution/node/208/log

daosbuild1 commented:
Test stage Build RPM on SLES 12.3 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1048/29/execution/node/235/log

brianjmurrell commented on ftest.sh:
```diff
@@ -125,8 +125,10 @@ if ${TEARDOWN_ONLY:-false}; then
-mkdir -p src/tests/ftest/avocado/job-results
+rm -rf install/lib/daos/TESTING/ftest/avocado ./*_results.xml
+#rm -rf src/tests/ftest/avocado ./*_results.xml
+mkdir -p install/lib/daos/TESTING/ftest/avocado/job-results
```
This is fine for this stage of the migration but in the final result, this needs to be portable to both the `install/` prefix and the RPM `/usr` prefix.

Additionally, in the latter case, this will not be writable so this needs to be somewhere else completely in such a case.  A "tempdir" of some sort seems to be the solution.

GitHuaKuang replied:
Tested adding TMPDIR in .build_vars.json worked for non-CI use case.

GitHuaKuang replied:
Use TMPDIR env variable if set, otherwise, default it to PREFIX

brianjmurrell replied:
@GitHuaKuang I've **Unresolved** this conversation.  Generally, it is up to the person who made the original comment to resolve a conversation as they need to verify that their concern has been addressed.

As for this concern, I'm not seeing any code change which addresses the concern.  The code as it is here will fail to work when a test run from RPMs is done.

GitHuaKuang replied:
the code is in ftest/util/server_utils.py

brianjmurrell replied:
This appears to be unresolved still.  In general, any hard-coded path under `install/` in this file will not work portably between RPM-based installations and non-RPM installations.  That goes for (just about) all instances of the use of `install/` in this file.

daosbuild1 commented:
Test stage Test SLES12.3 RPMs completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1048/32/execution/node/516/log

daosbuild1 commented:
Test stage Test CentOS 7 RPMs completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1048/32/execution/node/552/log

daosbuild1 commented:
Test stage Test SLES12.3 RPMs completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1048/33/execution/node/508/log

daosbuild1 commented:
Test stage Build RPM on CentOS 7 completed with status UNSTABLE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1048/34/execution/node/200/log

daosbuild1 commented:
Test stage Build RPM on SLES 12.3 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1048/34/execution/node/235/log

daosbuild1 commented:
Test stage Test SLES12.3 RPMs completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1048/36/execution/node/546/log

brianjmurrell commented on ftest.sh:
```diff
@@ -294,7 +296,8 @@ wq
 fi
 
-pushd src/tests/ftest
+#pushd src/tests/ftest
```
I would **strongly** suggest you move this dir to a new name (or just `rm -rf` it even) to be sure that nothing you don't know about (yet) is still referencing it or working out of it.

GitHuaKuang replied:
Tested with this suggestion, moved this dir to a new one before the test, and moved it back after. It passed all CI tests.

jolivier23 commented on Jenkinsfile:
```diff
@@ -1124,22 +1124,22 @@ pipeline {
                                            tnodes=$(echo $NODELIST | cut -d ',' -f 1-9)
                                            ./ftest.sh "$test_tag" $tnodes''',
-                                junit_files: "src/tests/ftest/avocado/*/*/*.xml src/tests/ftest/*_results.xml",
+                                junit_files: "install/lib/daos/TESTING/ftest/avocado/*/*/*.xml install/lib/daos/TESTING/ftest/*_results.xml",
```
Why don't we put tests in lib64 if we are placing libraries there?

brianjmurrell replied:
My only thought was that they are ultimately not 64-bit-only.  That said, I don't have a strong feeling one way or the other.

Ultimately these should be packaged up as a first-class python package i think.  But that can happen down the road and after I have given it more thought.

GitHuaKuang replied:
We will keep this approach in this PR, can have further discussion on other concerns.

jolivier23 commented on src/bio/SConscript:
```diff
@@ -16,7 +16,7 @@ def scons():
     bio = daos_build.library(denv, "bio", Glob('*.c'),
                              LIBS=['numa', 'spdk', 'smd'])
-    denv.Install('$PREFIX/lib/daos_srv', bio)
+    denv.Install('$PREFIX/lib64/daos_srv', bio)
```
It won't let me comment on the scons_local change...was it intentional?   If not, please revert it.   I can't tell what changed but scons_local was updated recently in master.

GitHuaKuang replied:
What's the best way to revert this scons_local commit accidentally made in solving conflict in master merge.
 Submodule scons_local updated 1 files
+21 −14		components/__init__.py

Do I have to revert all 2 commits + 1 merge in between, or is there a way to just update this __init__.py to master version?

GitHuaKuang replied:
With help from Jeff, this is resolved.

jolivier23 commented on SConstruct:
```diff
@@ -279,9 +279,13 @@ def scons():
-    env.InstallAs("$PREFIX/TESTING/.build_vars.sh", ".build_vars.sh")
-    env.InstallAs("$PREFIX/TESTING/.build_vars.json", ".build_vars.json")
-    env.InstallAs("$PREFIX/lib/daos/VERSION", "VERSION")
+    env.InstallAs("$PREFIX/lib/daos/TESTING/.build_vars.sh", ".build_vars.sh")
```
You don't need to use InstallAs unless you are changing the name of the file....so this could simply be 

env.Install("$PREFIX/lib/daos/TESTING", ".build_vars.sh")
same for other lines.

This type of code could benefit from a function, maybe in daos_build

daos_build.install(env, "lib64/daos", file)

where said function is
def install(env, subdir, files): #files can be string or an array
    """install files in standard location
    path = "$PREFIX/%s" % subdir
    env.Install(path, files)

Not really your fault but I don't like repetition.

jolivier23 commented on utils/rpms/daos.spec:
```diff
@@ -130,19 +130,21 @@ scons %{?no_smp_mflags}              \
-cp -al src/utils/py/daos_cref.py %{?buildroot}%{daoshome}/utils/py
-cp -al src/utils/py/conversion.py %{?buildroot}%{daoshome}/utils/py
+#mkdir -p %{?buildroot}%{daoshome}/utils/py
+#cp -al src/utils/py/daos_api.py %{?buildroot}%{daoshome}/utils/py
```
Is this a temporary change?  If not, just remove the lines instead of commenting.   We can tell what changed in the diff.

GitHuaKuang replied:
yes, will remove those commented lines since this PR passed all CI tests now.

daosbuild1 commented:
Test stage Functional completed with status FAILURE.  https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-1048/39/display/redirect

daosbuild1 commented:
Test stage Functional_Hardware completed with status FAILURE.  https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-1048/39/display/redirect

daosbuild1 commented:
Test stage Build RPM on SLES 12.3 completed with status UNSTABLE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1048/44/execution/node/265/log

daosbuild1 commented:
Test stage Build RPM on Leap 42.3 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1048/45/execution/node/256/log

brianjmurrell commented on src/tests/ftest/launch.py:
```diff
@@ -108,8 +108,8 @@ def set_test_environment():
-        os.path.join(base_dir, "lib", "python2.7", "site-packages"),
-        os.path.join(base_dir, "lib", "python3", "site-packages"),
+        os.path.join(base_dir, "lib64", "python2.7", "site-packages"),
+        os.path.join(base_dir, "lib64", "python3", "site-packages"),
```
Always dding both of these paths to the `PYTHONPATH` feels dangerous.  If the same module is available in both, won't the 2.7 module get used even when using Python 3?

brianjmurrell commented on src/utils/SConscript:
```diff
@@ -21,5 +22,9 @@ def scons():
     denv.Install('$PREFIX/bin/', dcont)
     denv.Install('$PREFIX/bin/', daos)
 
+    # copy utils/py to $PREFIX/lib/daos/utils/py
```
Shouldn't this to into a `site-packages` python dir?

brianjmurrell commented on utils/rpms/daos.spec:
```diff
@@ -157,19 +157,10 @@ scons %{?no_smp_mflags}              \
-cp -al src/utils/py/daos_cref.py %{?buildroot}%{daoshome}/utils/py
-cp -al src/utils/py/conversion.py %{?buildroot}%{daoshome}/utils/py
+sed -i -e s/${BUILDROOT//\//\\/}[^\"]\*/${PREFIX//\//\\/}/g %{?buildroot}%{daoshome}/TESTING/.build_vars.*
+sed -i -e s/${BUILDROOT//\//\\/}[^\"]\*/${PREFIX//\//\\/}/g %{?buildroot}%{_prefix}/lib/.build_vars.*
```
Why do we have two sets of `.build_vars.*` files?

GitHuaKuang replied:
The location $PREFIX/TESTING is generated by cart, not sure if it's safe to make any changes there:
./_build.external/cart/SConstruct:    env.InstallAs('$PREFIX/TESTING/.build_vars.sh', sh_build_vars)
./_build.external/cart/SConstruct:    env.InstallAs('$PREFIX/TESTING/.build_vars.json', json_build_vars)

brianjmurrell replied:
When DAOS is being built as an RPM, cart is not built so there shouldn't be anything under `install/` from cart.  You are probably confusing what gets put into `install/` during an RPM build vs. a full daos-stack build.

brianjmurrell commented on utils/rpms/daos.spec:
```diff
@@ -157,19 +157,10 @@ scons %{?no_smp_mflags}              \
+sed -i -e s/${BUILDROOT//\//\\/}[^\"]\*/${PREFIX//\//\\/}/g %{?buildroot}%{daoshome}/TESTING/.build_vars.*
+sed -i -e s/${BUILDROOT//\//\\/}[^\"]\*/${PREFIX//\//\\/}/g %{?buildroot}%{_prefix}/lib/.build_vars.*
+mkdir -p %{?buildroot}/%{_exec_prefix}/lib64/%{name}
+mkdir -p %{?buildroot}/%{_exec_prefix}/lib64/%{name}/certgen
```
Why do these two directories need to be created here?

brianjmurrell commented on SConstruct:
```diff
@@ -279,9 +280,11 @@ def scons():
+    daos_build.install(env, "lib/daos/TESTING/",
+                       ['.build_vars.sh', '.build_vars.json'])
+    # also install to $PREFIX/lib to work with existing avocado test code
+    daos_build.install(env, "lib/", ['.build_vars.sh', '.build_vars.json'])
```
Why are these `.build_vars.*` files being installed in more than one location?

GitHuaKuang replied:
Same reason: needed by cart.

daosbuild1 commented on utils/rpms/daos.spec:
```diff
@@ -264,6 +257,13 @@ install -m 644 utils/systemd/daos-agent.service %{?buildroot}/%{_unitdir}
 %{_bindir}/daos_gen_io_conf
 %{_bindir}/daos_run_io_conf
+# For avocado tests
+%dir %{_prefix}/lib 
```
(style)  trailing whitespace

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1048/46/execution/node/47/log

daosbuild1 commented:
Test stage Build RPM on SLES 12.3 completed with status UNSTABLE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1048/46/execution/node/229/log

daosbuild1 commented:
Test stage Build RPM on SLES 12.3 completed with status UNSTABLE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1048/47/execution/node/240/log

daosbuild1 commented:
Test stage Build RPM on CentOS 7 completed with status UNSTABLE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1048/47/execution/node/254/log

daosbuild1 commented:
Test stage Build RPM on Leap 42.3 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1048/47/execution/node/253/log

daosbuild1 commented:
Test stage Build RPM on SLES 12.3 completed with status UNSTABLE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1048/48/execution/node/243/log

daosbuild1 commented:
Test stage run_test.sh completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1048/49/execution/node/395/log

daosbuild1 commented:
Test stage run_test.sh completed with status FAILURE.  https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-1048/50/display/redirect

daosbuild1 commented:
Test stage Test CentOS 7 RPMs completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1048/51/execution/node/490/log

daosbuild1 commented:
Test stage Functional_Hardware completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1048/52/execution/node/532/log

daosbuild1 commented:
Test stage Functional_Hardware completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1048/53/execution/node/248/log

daosbuild1 commented:
Test stage Functional_Hardware completed with status FAILURE.  https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-1048/54/display/redirect

daosbuild1 commented on ftest.sh:
```diff
@@ -125,14 +125,18 @@ if ${TEARDOWN_ONLY:-false}; then
 # let's output to a dir in the tree
-rm -rf src/tests/ftest/avocado ./*_results.xml
-mkdir -p src/tests/ftest/avocado/job-results
+rm -rf $SL_PREFIX/lib/daos/TESTING/ftest/avocado ./*_results.xml
```
(lint) Double quote to prevent globbing and word splitting. [SC2086]

daosbuild1 commented on ftest.sh:
```diff
@@ -125,14 +125,18 @@ if ${TEARDOWN_ONLY:-false}; then
-rm -rf src/tests/ftest/avocado ./*_results.xml
-mkdir -p src/tests/ftest/avocado/job-results
+rm -rf $SL_PREFIX/lib/daos/TESTING/ftest/avocado ./*_results.xml
+mkdir -p $SL_PREFIX/lib/daos/TESTING/ftest/avocado/job-results
```
(lint) Double quote to prevent globbing and word splitting. [SC2086]

daosbuild1 commented on ftest.sh:
```diff
@@ -192,8 +196,12 @@ args+=" $*"
 ulimit -c unlimited
-rm -rf $DAOS_BASE/install/tmp
-mkdir -p $DAOS_BASE/install/tmp
+if [ $SL_PREFIX == "/usr" ]; then
```
(lint) The double quotes around this do nothing. Remove or escape them. [SC2140]

daosbuild1 commented on ftest.sh:
```diff
@@ -125,14 +125,18 @@ if ${TEARDOWN_ONLY:-false}; then
 
-DAOS_BASE=${SL_PREFIX%/install}
+if [ $SL_PREFIX == "/usr" ]; then
+    DAOS_BASE = '/'
```
(lint) Don't put spaces around the = in assignments. [SC1068]

daosbuild1 commented on ftest.sh:
```diff
@@ -125,14 +125,18 @@ if ${TEARDOWN_ONLY:-false}; then
 CLUSH_ARGS=($CLUSH_ARGS)
 
-DAOS_BASE=${SL_PREFIX%/install}
+if [ $SL_PREFIX == "/usr" ]; then
```
(lint) Double quote to prevent globbing and word splitting. [SC2086]

daosbuild1 commented on ftest.sh:
```diff
@@ -320,8 +332,11 @@ fi
 # jenkins build artifacts
-unlink $DAOS_BASE/src/tests/ftest/avocado/job-results/latest
-
+if [ $SL_PREFIX == "/usr" ]; then
```
(lint) The double quotes around this do nothing. Remove or escape them. [SC2140]

daosbuild1 commented on ftest.sh:
```diff
@@ -205,7 +213,11 @@ export D_LOG_FILE=\"$TEST_TAG_DIR/daos.log\"
 cat <<EOF > ~/.config/avocado/avocado.conf
 [datadir.paths]
-logs_dir = $DAOS_BASE/src/tests/ftest/avocado/job-results
+if [ $SL_PREFIX == "/usr" ]; then
```
(lint) The double quotes around this do nothing. Remove or escape them. [SC2140]

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1048/56/execution/node/47/log

daosbuild1 commented:
Test stage Functional completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1048/56/execution/node/519/log

daosbuild1 commented on ftest.sh:
```diff
@@ -320,8 +341,11 @@ fi
 # jenkins build artifacts
-unlink $DAOS_BASE/src/tests/ftest/avocado/job-results/latest
-
+if [ $SL_PREFIX == "/usr" ]; then
```
(lint) The double quotes around this do nothing. Remove or escape them. [SC2140]

daosbuild1 commented on ftest.sh:
```diff
@@ -125,14 +126,23 @@ if ${TEARDOWN_ONLY:-false}; then
-rm -rf src/tests/ftest/avocado ./*_results.xml
-mkdir -p src/tests/ftest/avocado/job-results
-
+if [ $SL_PREFIX == /usr ]; then
```
(lint) Double quote to prevent globbing and word splitting. [SC2086]

daosbuild1 commented on ftest.sh:
```diff
@@ -125,14 +126,23 @@ if ${TEARDOWN_ONLY:-false}; then
 CLUSH_ARGS=($CLUSH_ARGS)
 
-DAOS_BASE=${SL_PREFIX%/install}
+if [ $SL_PREFIX == /usr ]; then
```
(lint) Double quote to prevent globbing and word splitting. [SC2086]

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1048/57/execution/node/47/log

daosbuild1 commented:
Test stage run_test.sh completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1048/59/execution/node/230/log

brianjmurrell commented on utils/run_test.sh:
```diff
@@ -98,9 +98,10 @@ if [ -d "/mnt/daos" ]; then
     # Satisfy CGO requirements for go-spdk binding and internal daos imports
-    LD_LIBRARY_PATH="${SL_PREFIX}/lib:${SL_SPDK_PREFIX}/lib:${LD_LIBRARY_PATH}"
+    LD_LIBRARY_PATH=
+    "${SL_PREFIX}/lib64:${SL_SPDK_PREFIX}/lib:${LD_LIBRARY_PATH}"
```
You can't put these on a separate line, at least not without a `\` at the end of the first line.  I.e change the above two lines to:

```suggestion
    LD_LIBRARY_PATH=\
    "${SL_PREFIX}/lib64:${SL_SPDK_PREFIX}/lib:${LD_LIBRARY_PATH}"
```

daosbuild1 commented on utils/run_test.sh:
```diff
@@ -98,9 +98,10 @@ if [ -d "/mnt/daos" ]; then
     run_test src/rdb/raft_tests/raft_tests.py
     # Satisfy CGO requirements for go-spdk binding and internal daos imports
-    LD_LIBRARY_PATH="${SL_PREFIX}/lib:${SL_SPDK_PREFIX}/lib:${LD_LIBRARY_PATH}"
+    LD_LIBRARY_PATH=\
```
(lint) This assignment is only seen by the forked process. [SC2097]
(lint) Remove space after = if trying to assign a value (for empty string, use var='' ... ). [SC1007]

brianjmurrell replied:
Maybe you can't split an assignment like this.

daosbuild1 commented on utils/run_test.sh:
```diff
@@ -98,9 +98,10 @@ if [ -d "/mnt/daos" ]; then
     # Satisfy CGO requirements for go-spdk binding and internal daos imports
-    LD_LIBRARY_PATH="${SL_PREFIX}/lib:${SL_SPDK_PREFIX}/lib:${LD_LIBRARY_PATH}"
+    LD_LIBRARY_PATH=\
+    "${SL_PREFIX}/lib64:${SL_SPDK_PREFIX}/lib:${LD_LIBRARY_PATH}"
```
(lint) This expansion will not see the mentioned assignment. [SC2098]

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1048/60/execution/node/47/log

daosbuild1 commented:
Test stage Functional_Hardware completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1048/61/execution/node/512/log

daosbuild1 commented:
Test stage Functional completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1048/62/execution/node/504/log

daosbuild1 commented:
Test stage Functional completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1048/63/execution/node/504/log

daosbuild1 commented:
Test stage Functional_Hardware completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1048/63/execution/node/555/log

brianjmurrell commented on Jenkinsfile:
```diff
@@ -1207,36 +1207,37 @@ pipeline {
                                       arts="$arts$(ls *daos{,_agent}.log* 2>/dev/null)" && arts="$arts"$'\n'
-                                      arts="$arts$(ls -d src/tests/ftest/avocado/job-results/* 2>/dev/null)" && arts="$arts"$'\n'
-                                      arts="$arts$(ls src/tests/ftest/*.stacktrace 2>/dev/null || true)"
+                                      arts="$arts$(ls -d /ftest/avocado/job-results/* 2>/dev/null)" && arts="$arts"$'\n'
```
```suggestion
                                      arts="$arts$(ls -d /tmp/ftest/avocado/job-results/* 2>/dev/null)" && arts="$arts"$'\n'
```
perhaps?

daosbuild1 commented:
Test stage Functional completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1048/64/execution/node/503/log

daosbuild1 commented:
Test stage Functional_Hardware completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1048/64/execution/node/554/log

daosbuild1 commented:
Test stage run_test.sh completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1048/65/execution/node/229/log

daosbuild1 commented:
Test stage run_test.sh completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1048/66/execution/node/155/log

daosbuild1 commented:
Test stage run_test.sh completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1048/67/execution/node/229/log

daosbuild1 commented:
Test stage Functional_Hardware completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1048/68/execution/node/520/log

daosbuild1 commented:
Test stage Functional completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1048/68/execution/node/554/log

daosbuild1 commented:
Test stage run_test.sh completed with status FAILURE.  https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-1048/69/display/redirect

daosbuild1 commented:
Test stage run_test.sh completed with status FAILURE.  https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-1048/70/display/redirect

daosbuild1 commented on src/tests/ftest/util/server_utils.py:
```diff
@@ -186,12 +191,13 @@ def run_server(hostfile, setname, basepath, uri_path=None, env_dict=None,
             server_cmd.extend(["-x", "PATH"])
 
+        # build shared dir for tmp amd yaml files
+        tmpdir = os.getenv('TMPDIR', os.path.join(build_vars["PREFIX"], 'tmp')) 
```
(style)  trailing whitespace

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1048/71/execution/node/47/log

daosbuild1 commented on src/tests/ftest/util/server_utils.py:
```diff
@@ -293,13 +292,17 @@ def run_server(test, hostfile, setname, uri_path=None, env_dict=None,
+        # build shared dir for tmp amd yaml files
+        tmpdir = os.getenv('DAOS_TEST_SHARED_DIR', os.path.join(build_vars["PREFIX"], 'tmp'))
+
+	# set env CRT_ATTACH_INFO_PATH 
```
(style)  trailing whitespace

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1048/74/execution/node/48/log

daosbuild1 commented:
Test stage Functional completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1048/76/execution/node/519/log

daosbuild1 commented:
Test stage Functional completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1048/77/execution/node/235/log

daosbuild1 commented:
Test stage Build RPM on CentOS 7 completed with status UNSTABLE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1048/78/execution/node/216/log

daosbuild1 commented:
Test stage Build on Leap 42.3 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1048/78/execution/node/224/log

daosbuild1 commented:
Test stage Build on SLES 12.3 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1048/78/execution/node/177/log

daosbuild1 commented:
Test stage Build on CentOS 7 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1048/78/execution/node/235/log

daosbuild1 commented:
Test stage Build on Ubuntu 18.04 with Clang completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1048/78/execution/node/227/log

daosbuild1 commented:
Test stage Build on Leap 15 with Intel-C completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1048/78/execution/node/221/log

daosbuild1 commented:
Test stage Build RPM on CentOS 7 completed with status UNSTABLE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1048/79/execution/node/209/log

daosbuild1 commented:
Test stage Test CentOS 7 RPMs completed with status FAILURE.  https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-1048/80/display/redirect

daosbuild1 commented:
Test stage Functional_Hardware completed with status FAILURE.  https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-1048/80/display/redirect

daosbuild1 commented:
Test stage run_test.sh completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1048/81/execution/node/395/log

daosbuild1 commented:
Test stage Functional completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1048/82/execution/node/550/log

daosbuild1 commented:
Test stage Functional_Hardware completed with status FAILURE.  https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-1048/83/display/redirect

daosbuild1 commented:
Test stage Functional_Hardware completed with status FAILURE.  https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-1048/84/display/redirect

daosbuild1 commented:
Test stage Functional completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1048/85/execution/node/230/log

daosbuild1 commented:
Test stage Functional completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1048/86/execution/node/243/log

daosbuild1 commented:
Test stage Functional completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1048/87/execution/node/220/log

phender commented on src/tests/ftest/launch.py:
```diff
@@ -26,6 +26,7 @@
 from argparse import ArgumentParser, RawDescriptionHelpFormatter
 import json
 from multiprocessing import Process
+import sys
```
This line should be removed.  There is no new code being added using `sys` and the existing code using `sys` is handled by line 34.

phender commented on src/tests/ftest/pool/bad_evict.py:
```diff
@@ -94,7 +94,7 @@ def test_evict(self):
             with open('../../../.build_vars.json') as build_file:
                 data = json.load(build_file)
-            context = DaosContext(data['PREFIX'] + '/lib/')
+            context = DaosContext(data['PREFIX'] + '/lib64/')
```
Since this test inherits TestWithServers, this code is already executed in `TestWithServers.setup()` and the `DaosContext` object is assigned to `self.context`.  That means the pool creation can simple be `pool = DaosPool(self.context)` and all the code using `json` (including the import) and assigning `context` can be removed.

To make use of the existing `TestWithServers.tearDown()` functionality the `pool` variable should be `self.pool`.

phender commented on src/tests/ftest/pool/bad_query.py:
```diff
@@ -81,7 +81,7 @@ def test_query(self):
             with open('../../../.build_vars.json') as build_file:
                 data = json.load(build_file)
-            context = DaosContext(data['PREFIX'] + '/lib/')
+            context = DaosContext(data['PREFIX'] + '/lib64/')
```
Since this test inherits TestWithServers, this code is already executed in `TestWithServers.setup()` and the `DaosContext` object is assigned to `self.context`.  That means the pool creation can simple be `pool = DaosPool(self.context)` and all the code using `json` (including the import) and assigning `context` can be removed.

To make use of the existing `TestWithServers.tearDown()` functionality the `pool` variable should be `self.pool`.

phender commented on src/tests/ftest/util/apricot/apricot/test.py:
```diff
@@ -121,23 +121,26 @@ def setUp(self):
         self.daosctl = os.path.join(self.prefix, 'bin', 'daosctl')
 
+        if self.prefix != "/usr":
+            self.tmp = os.path.join(build_paths["PREFIX"], 'tmp')
```
```suggestion
            self.tmp = os.path.join(self.prefix, 'tmp')
```

GitHuaKuang replied:
done

phender commented on src/tests/ftest/util/apricot/apricot/test.py:
```diff
@@ -121,23 +121,26 @@ def setUp(self):
+        if self.prefix != "/usr":
+            self.tmp = os.path.join(build_paths["PREFIX"], 'tmp')
+        else:
+            self.tmp = os.path.expanduser('~/daos_test')
```
Should this be `~/daos_test/tmp`?  Do you need to `mkdir -p` this directory?

phender commented on src/tests/ftest/util/server_utils.py:
```diff
@@ -241,10 +238,24 @@ def run_server(test, hostfile, setname, uri_path=None, env_dict=None,
+        else:
+            default_yamldir = build_vars["PREFIX"]
+
+        yaml_prefix = os.getenv('DAOS_TEST_SHARED_DIR', default_yamldir)
```
The `test` argument passed into this function is a `TestWithServers` object, so you can just use `test.tmp` in place of `default_yamldir`. e.g.
```
yaml_prefix = os.getenv('DAOS_TEST_SHARED_DIR', test.tmp)
```
or even the following if the assignment of `self.tmp` included the optioanl `DAOS_TEST_SHARED_DIR` env handling in the `TestWithServers.setUp()` method:
```
server_yaml = os.path.join(test.tmp, AVOCADO_FILE)
```

GitHuaKuang replied:
Got this error after using test.tmp:
08:44:43 DEBUG|   File "/var/lib/jenkins/jenkins-1/docker_1/workspace/daos-stack_daos_PR-1048/install/lib/daos/TESTING/ftest/util/server_utils.py", line 248, in run_server
08:44:43 DEBUG|     server_yaml = os.path.join(test.tmp, AVOCADO_FILE)
08:44:43 DEBUG|   File "/usr/lib64/python2.7/posixpath.py", line 77, in join
08:44:43 DEBUG|     elif path == '' or path.endswith('/'):
08:44:43 DEBUG| AttributeError: 'NoneType' object has no attribute 'endswith'

GitHuaKuang replied:
Is test.tmp the same as self.tmp set in test.py?
        if self.prefix != "/usr":
131             self.tmp = os.path.join(self.prefix, 'tmp')
132         else:
133             self.tmp = os.getenv('DAOS_TEST_SHARED_DIR', \
134                                  os.path.expanduser('~/daos_test'))
135             os.makedirs(self.tmp, exist_ok = True)

daosbuild1 commented:
Test stage Functional completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1048/88/execution/node/511/log

phender commented on src/tests/ftest/util/server_utils.py:
```diff
@@ -292,14 +299,21 @@ def run_server(test, hostfile, setname, uri_path=None, env_dict=None,
+            default_tmpdir = os.path.expanduser('~/daos_test')
+
+        # build shared dir for tmp amd yaml files
+        tmpdir = os.getenv('DAOS_TEST_SHARED_DIR', default_tmpdir)
```
This should already assigned in `test.tmp`.  See previous comment for more details.

GitHuaKuang replied:
done

phender commented on src/tests/ftest/util/server_utils.py:
```diff
@@ -269,10 +280,6 @@ def run_server(test, hostfile, setname, uri_path=None, env_dict=None,
-        # Pile of build time variables
-        with open(os.path.join(test.basepath, ".build_vars.json")) as json_vars:
-            build_vars = json.load(json_vars)
-
```
This is needed for more than just determining the shared tmp dir.  The subsequent lines will result in an error.  This should be restored.

phender commented on src/tests/ftest/util/server_utils.py:
```diff
@@ -292,14 +299,21 @@ def run_server(test, hostfile, setname, uri_path=None, env_dict=None,
+        tmpdir = os.getenv('DAOS_TEST_SHARED_DIR', default_tmpdir)
 
+        # set env CRT_ATTACH_INFO_PATH
+        os.environ["CRT_ATTACH_INFO_PATH"] = tmpdir
```
If we are dynamically assigning the `CRT_ATTACH_INFO_PATH` env, then it needs to be exported by the `orterun` command so the other nodes know it is located.

phender commented on src/utils/SConscript:
```diff
@@ -1,5 +1,6 @@
 """Build DAOS utilities"""
 import daos_build
+import os
```
Is this intentional?  Why are we adding an import and not making use of it?

GitHuaKuang replied:
done

phender commented on ftest.sh:
```diff
@@ -125,8 +125,8 @@ if ${TEARDOWN_ONLY:-false}; then
 # let's output to a dir in the tree
-rm -rf src/tests/ftest/avocado ./*_results.xml
-mkdir -p src/tests/ftest/avocado/job-results
+rm -rf install/lib/daos/TESTING/ftest/avocado ./*_results.xml
```
It would appear that assigning `install/lib/daos/TESTING` (or whatever ends up being used) to a variable would make sense given its reuse.

GitHuaKuang replied:
https://jira.hpdd.intel.com/browse/DAOS-3607
Will move CI test to DAOS RPM env.

daosbuild1 commented:
Test stage Functional completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1048/89/execution/node/511/log

GitHuaKuang commented on src/tests/ftest/pool/bad_create.py:
```diff
@@ -118,7 +118,7 @@ def test_create(self):
             with open('../../../.build_vars.json') as build_file:
                 data = json.load(build_file)
-            context = DaosContext(data['PREFIX'] + '/lib/')
+            context = DaosContext(data['PREFIX'] + '/lib64/')
```
done

daosbuild1 commented on src/tests/ftest/pool/global_handle.py:
```diff
@@ -81,6 +46,26 @@ def tearDown(self):
             check_for_pool.cleanup_pools(self.hostlist_servers)
 
+    def check_handle(self, buf_len, iov_len, buf, uuidstr, rank):
+        """ 
```
(style)  trailing whitespace

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1048/90/execution/node/48/log

daosbuild1 commented on src/tests/ftest/container/global_handle.py:
```diff
@@ -26,40 +26,47 @@
+        """
+        This gets run in a child process and verifyes the global
+        handles can be turned into local handles in another process.
+        """                             
```
(style)  trailing whitespace

daosbuild1 commented on src/tests/ftest/container/global_handle.py:
```diff
@@ -26,40 +26,47 @@
                           ctypes.POINTER(ctypes.c_byte *
                                          pool_glob_handle.iov_buf_len))
-        buf2 = bytearray()
+        buf2 = bytearray()                                           
```
(style)  trailing whitespace

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1048/92/execution/node/49/log

daosbuild1 commented:
Test stage Functional completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1048/93/execution/node/512/log

GitHuaKuang commented:
Can you please take one more check see if anything has to be changed before landing for 0.8 release? It passed all functional test yesterday with some minor changes added today, since the CI test may take quite a while, maybe we can do the review on parallel.

brianjmurrell commented on ftest.sh:
```diff
@@ -205,7 +205,7 @@ export D_LOG_FILE=\"$TEST_TAG_DIR/daos.log\"
 cat <<EOF > ~/.config/avocado/avocado.conf
 [datadir.paths]
-logs_dir = $DAOS_BASE/src/tests/ftest/avocado/job-results
+logs_dir = $DAOS_BASE/install/lib/daos/TESTING/ftest/avocado/job-results
```
Ditto.

GitHuaKuang replied:
https://jira.hpdd.intel.com/browse/DAOS-3607
Will move CI test to DAOS RPM env.

brianjmurrell commented on ftest.sh:
```diff
@@ -294,7 +294,7 @@ wq
 fi
 
-pushd src/tests/ftest
+pushd install/lib/daos/TESTING/ftest
```
Ditto.

GitHuaKuang replied:
This line change is the 1st step that moves CI test to use the newly located avocado tests, next step will be addressed by:
https://jira.hpdd.intel.com/browse/DAOS-3607
Will move CI test to DAOS RPM env.

brianjmurrell commented on ftest.sh:
```diff
@@ -313,7 +313,7 @@ fi
 # Remove the latest avocado symlink directory to avoid inclusion in the
 # jenkins build artifacts
-unlink $DAOS_BASE/src/tests/ftest/avocado/job-results/latest
+unlink $DAOS_BASE/install/lib/daos/TESTING/ftest/avocado/job-results/latest
```
Ditto.

brianjmurrell commented on src/tests/ftest/util/apricot/apricot/test.py:
```diff
@@ -121,23 +121,31 @@ def setUp(self):
         self.daosctl = os.path.join(self.prefix, 'bin', 'daosctl')
 
+        # set default shared dir for daos tests in case DAOS_TEST_SHARED_DIR$
+        # is not set, for RPM env and non-RPM env.$
```
Not sure why there are trailing `$` at the end of these comment lines.

GitHuaKuang replied:
It's a copy/paste issue we also noticed in some python test code changes: I normally do ":set list" in vi editor to make sure no space trailer, but forgot to unlist them before making copies. Will clear them here.

GitHuaKuang replied:
done

brianjmurrell commented on utils/rpms/daos.spec:
```diff
@@ -293,6 +284,9 @@ install -m 644 utils/systemd/daos-agent.service %{?buildroot}/%{_unitdir}
 %{_bindir}/daos_run_io_conf
+# For avocado tests
+%{_prefix}/lib/.build_vars.json
+%{_prefix}/lib/.build_vars.sh
```
Same as above, these files should not be installed into `/usr/lib/` but rather `/usr/lib/daos/`

GitHuaKuang replied:
ok, will move the location and update .py code accordingly.

GitHuaKuang replied:
done

GitHuaKuang commented on SConstruct:
```diff
@@ -283,9 +284,11 @@ def scons():
+    # daos_build.install(env, "lib/daos/TESTING/",
+    #                   ['.build_vars.sh', '.build_vars.json'])
+    # also install to $PREFIX/lib to work with existing avocado test code
+    daos_build.install(env, "lib/", ['.build_vars.sh', '.build_vars.json'])
```
installed to $PREFIX/lib/daos

daosbuild1 commented:
Test stage Build RPM on Leap 15 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1048/96/execution/node/244/log

daosbuild1 commented:
Test stage Build RPM on Leap 15 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1048/98/execution/node/248/log

GitHuaKuang commented:
Addressed most comments and opened DAOS-3607 to handle CI test on RPM. Local unit tested and worked well, CI all functional test passed, but failed IorSmall in Functional-Hardware, filed DAOS-3616.

jolivier23 commented on utils/check_modules.sh:
```diff
@@ -41,8 +41,8 @@ set -e
 fi
-PYTHONPATH=$PWD/utils:$PWD/src/tests/ftest/util/
-PYTHONPATH=$PYTHONPATH:$PWD/src/tests/ftest/util/apricot/
+PYTHONPATH=$PWD/utils:$PWD/install/lib/daos/TESTING/ftest/util/
```
should this be lib64 too or is it lib for python?

GitHuaKuang replied:
This should be lib, util for ftest, not libraries (that go to lib64).

brianjmurrell replied:
I am still on the fence about this one and so it's subject to change.

I don't believe there is actually anything 64-bit specific in there so lib64/daos just seems wrong.  But it's also pretty clear from RPM packaging macros that on 64-bit machines, they don't want you using /usr/lib for anything.

So it probably needs some more investigation as to what best/accepted practices are.
